src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -241,6 +253,7 @@ public override bool CanApplyChange(ApplyChangesKind feature)
                 case ApplyChangesKind.AddAdditionalDocument:
                 case ApplyChangesKind.RemoveAdditionalDocument:
                 case ApplyChangesKind.ChangeAdditionalDocument:
+                case ApplyChangesKind.ChangeCompilationOptions:
                 case ApplyChangesKind.ChangeParseOptions:
```

Comment:
at first I just blindly implemented this to be same as ChangeParseOptions, but after thinking about this, this seems wrong. If this returns false for ChangeParseOptions and CanApplyParseOptionChange returns true, CheckAllowedProjectChanges is still fine with it. Therefore CanApplyParseOptionChange(ChangeParseOptions) really seems to be asking "Can you change *any* parse options" rather than "Are you able to change *some* parse options". Having this return true actually means that this workspace is fine will any kinds of parse/compilation option changes and will likely crash if that is done.

Reply:
This looks like an existing bug.

Reply:
Regardless, it should be cleared up and well documented as to what CanApplyChange(ChangeParseOptions) means and how it interacts with CanApplyParseOptionChange.

Reply:
Yeah, this isn't the best corner of API design here. "When in Rome, do as the Romans do", seems good advice here.

Reply:
@jasonmalinowski Are you saying we should leave this bug in?

Reply:
Yeah, it's not worth holding up the other improvements you've made here to fix the API design that's already broken. :smile:

Reply:
The fix would only involve removing these 2 cases here (ApplyChangesKind.ChangeCompilationOptions and ApplyChangesKind.ChangeParseOptions). The check succeeds without these because of the specialized checks in the other methods. These being here make the checks pass on any kind of changes even though the methods that apply the changes assume otherwise.

Reply:
The problem is that some methods assume 1 possible API design and this method assumes a different one.

Reply:
Well, the intended contract from callers is they check CanApplyChange before doing anything else. Any fix that's calling CanApplyChange (and getting lucky that the change they're trying to apply is actually supported) would be broken if we touched this.

Reply:
@jasonmalinowski That's not the intended contract. We talked about that here: https://github.com/dotnet/roslyn/pull/25875#discussion_r178672379 Please comment in there why you think it should be option 1 after all.

Reply:
As I said, I prefer option 2 because it's a little misleading that CanApplyChange would return true and then you get an exception saying the change isn't supported. Remember that CanApplyParseOptionChange isn't public, so you have no way to know whether the change is really supported and whether you're going to get an exception or not that way.

Reply:
@jasonmalinowski Also, CanApplyChange returning true for compilation option changes has always meant "you can change any compilation options". With your suggestion, it would mean "no, I only support some options, and you can't even check which ones because the check for this is not public.". Also, the default value of that specific check is false so anyone who has returned true for compilation options now would be broken now.

Reply:
> Also, CanApplyChange returning true for compilation option changes has always meant "you can change any compilation options"

It has always meant "the workspace supports changing some options in some cases." Practically, the workspace can always reject for all sorts of silliness. For example, even if a workspace says it supports document modification, it might fail if the file is read only. References coming from NuGet can't be removed (...sometimes.) Project references can't always be added if the projects aren't compatible. CanApplyChange is simply intended as a first-order bit for "does the workspace support changing options at all", to exclude things like MiscellaneousFiles where of course the answer is no.

As mentioned in the other thread: we should just make CanApplyParseOptions public, which I think solves your problems. Either way, let's stop _this_ thread so we only have one thread to discuss rather than two in parallel.